crates/ra_lsp_server/src/main_loop/handlers.rs

```diff
@@ -652,9 +655,87 @@ pub fn handle_code_lens(
         }
     }
 
+    // Handle impls
+    lenses.extend(
+        world
+            .analysis()
+            .file_structure(file_id)
+            .into_iter()
+            .filter(|it| match it.kind {
+                SyntaxKind::TRAIT_DEF | SyntaxKind::STRUCT_DEF | SyntaxKind::ENUM_DEF => true,
+                _ => false,
+            })
+            .map(|it| {
+                let range = it.node_range.conv_with(&line_index);
+                let pos = range.start;
+                let lens_params =
+                    req::TextDocumentPositionParams::new(params.text_document.clone(), pos);
+                CodeLens {
+                    range,
+                    command: None,
+                    data: Some(to_value(CodeLensResolveData::Impls(lens_params)).unwrap()),
+                }
+            }),
+    );
+
     return Ok(Some(lenses));
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+enum CodeLensResolveData {
+    Impls(req::TextDocumentPositionParams),
+}
+
+pub fn handle_code_lens_resolve(world: ServerWorld, code_lens: CodeLens) -> Result<CodeLens> {
+    let data = code_lens.data.unwrap();
+    let resolve = serde_json::from_value(data)?;
+    match resolve {
+        Some(CodeLensResolveData::Impls(lens_params)) => {
+            let locations: Vec<Location> =
+                match handle_goto_implementation(world, lens_params.clone())? {
+                    Some(req::GotoDefinitionResponse::Scalar(loc)) => vec![loc],
+                    Some(req::GotoDefinitionResponse::Array(locs)) => locs,
+                    Some(req::GotoDefinitionResponse::Link(links)) => links
+                        .into_iter()
+                        .map(|link| Location::new(link.target_uri, link.target_selection_range))
+                        .collect(),
+                    _ => vec![],
+                };
+
+            let title = if locations.len() == 1 {
+                "1 implementation".into()
+            } else {
+                format!("{} implementations", locations.len())
+            };
+
+            return Ok(CodeLens {
+                range: code_lens.range,
+                command: Some(Command {
+                    title,
+                    command: "rust-analyzer.showReferences".into(),
```

Comment:
Can we use `editor.action.showReferences` directly here?

Reply:
No because `editor.action.showReferences` requires the vscode types and not the lsp types. It took me a while to figure that out. vscode-java also does it this way.

Reply:
Let's add a comment here?

Reply:
Done.